File: js/src/lang/codes.test.ts
// The named code sets are pinned to conformance/cases/languages.json, which the
// Python reference generates. Python, Go and JS each assert against the same
// file, so the 41 codes cannot diverge by review error — the same three-way pin
// the ADR-0011 script CLAIM already has in tokenize-v2.test.ts.
//
// The compatibility half matters as much as the parity half: a plain string is
// assignable to every public parameter type here, forever. `enum` would have
// made that a type error, which is why this port uses a const object.
import { describe, expect, it } from "vitest";

import { loadCase } from "../conform/fixtures.js";
import { AUTO_ANSWER_LANGUAGE } from "./lang.js";
import {
  Language,
  type LanguageLike,
  SEARCH_LANGUAGES,
  Script,
  type ScriptLike,
  searchLanguageByCode,
} from "./codes.js";

interface LanguagesFixture {
  auto_sentinel: string;
  scripts: string[];
  supported_scripts: string[];
  continuous_scripts: string[];
  languages: { code: string; name: string; scripts: string[]; supported: boolean }[];
}

const FIXTURE = loadCase<LanguagesFixture>("languages.json");

/** Bucket sizes, pinned. A vector silently dropped from a bucket is a weakened
 *  contract that no per-case assertion can see. (`auto_sentinel` is a scalar,
 *  not a bucket, and is asserted by name below.) */
const EXPECTED_COUNTS: Record<string, number> = {
  scripts: 27,
  supported_scripts: 14,
  continuous_scripts: 7,
  languages: 41,
};

describe("languages.json bucket shape", () => {
  it("bucket names and sizes are pinned", () => {
    const sizes = Object.fromEntries(
      Object.entries(FIXTURE)
        .filter(([, v]) => Array.isArray(v))
        .map(([k, v]) => [k, (v as unknown[]).length]),
    );
    expect(sizes).toEqual(EXPECTED_COUNTS);
  });
});

describe("language / script code sets (conformance)", () => {
  it("the search table matches the fixture, in order", () => {
    expect(
      SEARCH_LANGUAGES.map((l) => ({
        code: l.code,
        name: l.name,
        scripts: [...l.scripts],
        supported: l.supported,
      })),
    ).toEqual(FIXTURE.languages);
  });

  it("the Language members are the search codes plus the auto sentinel", () => {
    const members = new Set<string>(Object.values(Language));
    const codes = new Set(FIXTURE.languages.map((l) => l.code));
    codes.add(FIXTURE.auto_sentinel);
    expect(members).toEqual(codes);
    expect(FIXTURE.languages).toHaveLength(41);
  });

  it("the Script members match the fixture", () => {
    expect([...Object.values(Script)].sort()).toEqual([...FIXTURE.scripts].sort());
  });

  it("the auto sentinel is named but is not searchable", () => {
    expect(Language.AUTO).toBe(FIXTURE.auto_sentinel);
    expect(AUTO_ANSWER_LANGUAGE).toBe(Language.AUTO);
    expect(searchLanguageByCode(Language.AUTO)).toBeUndefined();
  });
});

describe("plain strings stay first-class", () => {
  it("a bare string is assignable to LanguageLike and ScriptLike", () => {
    // The compile-time half of the pin: if these ever become a TS `enum`, the
    // next three lines stop type-checking and `npm run typecheck` fails.
    const fromLiteral: LanguageLike = "ta";
    const computed: LanguageLike = ["t", "a"].join("");
    const script: ScriptLike = "tamil";
    expect(fromLiteral).toBe(Language.TAMIL);
    expect(computed).toBe(Language.TAMIL);
    expect(script).toBe(Script.TAMIL);
  });

  it("lookup by a plain string and by a member are the same entry", () => {
    expect(searchLanguageByCode("ta")).toBe(searchLanguageByCode(Language.TAMIL));
    expect(searchLanguageByCode("ta")?.name).toBe("Tamil");
  });

  it("a typo resolves to nothing — codes are never guessed", () => {
    expect(searchLanguageByCode("tamiil")).toBeUndefined();
  });

  it("the answer-language chain takes a string or a member identically", () => {
    // resolveAnswerLanguage is the pinned §11a function; both forms are one call.
    expect(Language.TAMIL === "ta").toBe(true);
    expect(JSON.stringify({ answer_language: Language.TAMIL })).toBe('{"answer_language":"ta"}');
  });
});

describe("the refusal set is carried, not dropped", () => {
  it("names the languages it cannot search", () => {
    const unsupported = SEARCH_LANGUAGES.filter((l) => !l.supported).map((l) => l.code);
    expect(unsupported).toContain(Language.KANNADA);
    // Telugu IS claimed (it has a fixture); its Indic neighbours are not.
    expect(SEARCH_LANGUAGES.find((l) => l.code === Language.TELUGU)?.supported).toBe(true);
  });

  it("every script the table names is a declared Script", () => {
    const declared = new Set<string>(Object.values(Script));
    for (const l of SEARCH_LANGUAGES) {
      for (const s of l.scripts) expect(declared.has(s)).toBe(true);
    }
  });
});

File: golang/lang/codes_test.go
package lang

import (
	"reflect"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// The named code sets are pinned to conformance/cases/languages.json, which the
// Python reference generates. Python, Go and JS each assert against the same
// file, so the 41 codes cannot diverge by review error — the same three-way pin
// the ADR-0011 script CLAIM already has in tokenize_v2_test.go.

type languagesFixture struct {
	AutoSentinel      string   `json:"auto_sentinel"`
	Scripts           []string `json:"scripts"`
	SupportedScripts  []string `json:"supported_scripts"`
	ContinuousScripts []string `json:"continuous_scripts"`
	Languages         []struct {
		Code      string   `json:"code"`
		Name      string   `json:"name"`
		Scripts   []string `json:"scripts"`
		Supported bool     `json:"supported"`
	} `json:"languages"`
}

func loadLanguages(t *testing.T) languagesFixture {
	t.Helper()
	var f languagesFixture
	conform.Case(t, "languages.json", &f)
	return f
}

// expectedLanguagesCounts pins every list in conformance/cases/languages.json.
// The row-by-row loop below compares the table against the fixture, so a fixture
// that lost rows would agree with a table that lost the same rows unless the
// absolute sizes are pinned too.
var expectedLanguagesCounts = map[string]int{
	"scripts":            27,
	"supported_scripts":  14,
	"continuous_scripts": 7,
	"languages":          41,
}

func TestLanguagesVectorBucketSizes(t *testing.T) {
	f := loadLanguages(t)
	for name, got := range map[string]int{
		"scripts":            len(f.Scripts),
		"supported_scripts":  len(f.SupportedScripts),
		"continuous_scripts": len(f.ContinuousScripts),
		"languages":          len(f.Languages),
	} {
		if want := expectedLanguagesCounts[name]; got != want {
			t.Errorf("languages.json bucket %q: got %d entries, want %d", name, got, want)
		}
	}
}

func TestSearchLanguageTableMatchesConformance(t *testing.T) {
	f := loadLanguages(t)
	got := SearchLanguages()
	if len(got) != len(f.Languages) {
		t.Fatalf("table has %d languages, fixture has %d", len(got), len(f.Languages))
	}
	for i, want := range f.Languages {
		g := got[i]
		if string(g.Code) != want.Code || g.Name != want.Name || g.Supported != want.Supported {
			t.Errorf("row %d = {%s %s %v}, want {%s %s %v}",
				i, g.Code, g.Name, g.Supported, want.Code, want.Name, want.Supported)
		}
		scripts := make([]string, len(g.Scripts))
		for j, s := range g.Scripts {
			scripts[j] = string(s)
		}
		if !reflect.DeepEqual(scripts, want.Scripts) {
			t.Errorf("row %d scripts = %v, want %v", i, scripts, want.Scripts)
		}
	}
}

func TestAutoSentinelMatchesConformance(t *testing.T) {
	f := loadLanguages(t)
	if f.AutoSentinel != string(Auto) {
		t.Fatalf("Auto = %q, want %q", Auto, f.AutoSentinel)
	}
	if f.AutoSentinel != AutoAnswerLanguage {
		t.Fatalf("AutoAnswerLanguage = %q, want %q", AutoAnswerLanguage, f.AutoSentinel)
	}
	// The sentinel is NOT a searchable language.
	if _, ok := SearchLanguageByCode(Auto); ok {
		t.Fatal("the auto sentinel must not be in the search-language table")
	}
}

// TestPlainStringsStillAssign is the compatibility pin: Language and Script are
// DEFINED STRING TYPES, so an untyped constant converts implicitly and every
// pre-existing literal call site keeps compiling. If someone ever "improves"
// these into an iota enum, this stops building.
func TestPlainStringsStillAssign(t *testing.T) {
	var l Language = "ta"
	var s Script = "tamil"
	if l != Tamil || s != ScriptTamil {
		t.Fatalf("literal assignment diverged: %q %q", l, s)
	}
	if got := ResolveAnswerLanguage(nil, "ta", "", nil, "en"); got != "ta" {
		t.Fatalf("string answer language = %q, want %q", got, "ta")
	}
	if got := ResolveAnswerLanguage(nil, string(Tamil), "", nil, string(English)); got != string(Tamil) {
		t.Fatalf("member answer language = %q, want %q", got, Tamil)
	}
}

func TestLookupByCodeAcceptsAPlainString(t *testing.T) {
	byString, ok := SearchLanguageByCode("ta")
	if !ok {
		t.Fatal(`SearchLanguageByCode("ta") not found`)
	}
	byMember, ok := SearchLanguageByCode(Tamil)
	if !ok {
		t.Fatal("SearchLanguageByCode(Tamil) not found")
	}
	if !reflect.DeepEqual(byString, byMember) {
		t.Fatalf("string and member lookups differ: %+v vs %+v", byString, byMember)
	}
	if _, ok := SearchLanguageByCode("tamiil"); ok {
		t.Fatal("a typo must not resolve — codes are never guessed")
	}
}

// Every script the table names must be a declared Script constant, and every
// declared Script must appear in the fixture's script list.
func TestScriptConstantsMatchConformance(t *testing.T) {
	f := loadLanguages(t)
	declared := map[string]bool{}
	for _, s := range f.Scripts {
		declared[s] = true
	}
	for _, l := range SearchLanguages() {
		for _, s := range l.Scripts {
			if !declared[string(s)] {
				t.Errorf("%s names script %q, absent from the fixture", l.Code, s)
			}
		}
	}
	if len(f.Scripts) == 0 || !declared[string(ScriptLatin)] || !declared[string(ScriptUnknown)] {
		t.Fatal("fixture script list is missing latin/unknown")
	}
}
